### PR TITLE
Add PodTemplate semantic validation for Elasticsearch

### LINKED ISSUE #2266: Reconciliation is blocked when a Pod can't be created
I got a situation where:
1. Pod could not be created because of some invalid values in the spec
2. `ok, err := d.expectationsSatisfied()` always returns `false`
3. It is impossible for the user to recover from that situation

The following manifest helps to recreate the issue:

```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1beta1
kind: Elasticsearch
metadata:
  name: blocked
spec:
  version: 7.5.0
  nodeSets:
    - name: default
      count: 1
      config:
        node.store.allow_mmap: false
      podTemplate:
        spec:
          volumes:
            - name: BaDVolumeName
              emptyDir: {}
```

The problem is that we assume that a Pod is at least in a `Pending` state, which is not the case.

### PR DESCRIPTION
**Pull request by barkbay**
This PR adds a semantic validation for the content of the `podTemplate` field of the Elasticsearch resources.

It is achieved by performing a dry-run request which attempts to create a Pod reusing the content of the `podTemplate`. The validation is only performed when a `StatefulSet` is created or updated.

Fixes #2266

### CODE REVIEW

**Comment by barkbay** [comment]
The validation fails with the following error on GKE 1.14: `Internal error occurred: admission webhook "pod-ready.common-webhooks.networking.gke.io" does not support dry run`

I think it means that the validation proposed here should only be a "best effort" validation: as long as the error is not explicitly a `metav1.StatusReasonInvalid` (422) we should not block the reconciliation loop.

File: pkg/controller/elasticsearch/sset/validation.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	return fmt.Sprintf(
+		"Validation of PodTemplate for StatefulSet %s in Elasticsearch %s/%s failed for the following reasons: %v",
+		e.StatefulSet.Name,
+		e.Parent.GetNamespace(),
+		e.Parent.GetName(),
+		e.Causes,
+	)
+}
+
+// validatePodTemplate validates a Pod Template by issuing a dry API request.
```
Comment (sebgl) [inline pkg/controller/elasticsearch/sset/validation.go:36]: dry -> dry-run?

File: test/e2e/es/podtemplate_test.go [hunk 2]
```diff
@@ -7,20 +7,110 @@ package es
...
+// TestPodTemplateValidation simulates a cluster update with a temporary invalid pod template.
+// The invalid pod template should not prevent the update to be eventually applied.
+func TestPodTemplateValidation(t *testing.T) {
+
+	kubernetesVersion, err := semver.ParseTolerant(test.Ctx().KubernetesVersion)
+	if err != nil {
+		t.Fatalf(
+			"Failed to parse the Kubernetes version: %v, err is %v", test.Ctx().KubernetesVersion, err)
+	}
+	if kubernetesVersion.LT(semver.MustParse("1.15.0")) {
```
Comment (sebgl) [inline test/e2e/es/podtemplate_test.go:37]:
```
// Pod creation dry run is only supported starting 1.15
```

File: pkg/controller/elasticsearch/sset/validation.go (hunk 1 above)
Comment (sebgl) [reply pkg/controller/elasticsearch/sset/validation.go:36]: `// This check is performed as "best-effort", since the dry-run API is only available starting K8s 1.15.0`

File: test/e2e/es/podtemplate_test.go [hunk 3]
```diff
@@ -7,20 +7,110 @@ package es
...
+								},
+							}),
+						)
+					},
+				}).
+			WithSteps(
+				// Mutate Elasticsearch with an invalid PodTemplate
+				b.UpgradeTestSteps(k),
+			).WithStep(
+			test.Step{
```
Comment (sebgl) [inline test/e2e/es/podtemplate_test.go:74]: indentation feels a bit off here?

File: pkg/controller/elasticsearch/sset/validation.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+// toPodTemplateError attempts to extract the meaningful information from the dry run API call by converting the original
+// error into a PodTemplateError.
+func toPodTemplateError(
+	parent metav1.Object,
+	sset appsv1.StatefulSet,
+	err error,
+) error {
+	var statusErr *k8serrors.StatusError
+	if !errors.As(err, &statusErr) {
+		// Not a Kubernetes API error (e.g. timeout)
```
Comment (sebgl) [inline pkg/controller/elasticsearch/sset/validation.go:68]: I'm wondering if we could always reach a timeout in some situations (eg. particular proxy configuration). Hard to know for sure, so I'm fine moving on with this.
Should we add some debug logs here just in case?

**Comment by barkbay** [comment]
To explain a little bit more the issue with the admission webhook: as explained in the [documentation](https://kubernetes.io/docs/reference/using-api/api-concepts/#make-a-dry-run-request):

>  If the request would trigger an admission controller which would have side effects, the request will be failed rather than risk an unwanted side effect. [...] admission webhooks can declare in their configuration object that they do not have side effects by setting the sideEffects field to “None”. If a webhook actually does have side effects, then the sideEffects field should be set to “NoneOnDryRun”

In order for an admission webhook to state if it supports dry-run its `sideEffects` field must be set accordingly. If this field is not set the default value is `Unknown`. In this case the request fails with something like this: `admission webhook "iam-for-pods.amazonaws.com" does not support dry run`

```
$ k get MutatingWebhookConfiguration -o yaml
```
```yaml
apiVersion: v1
items:
- apiVersion: admissionregistration.k8s.io/v1
  kind: MutatingWebhookConfiguration
  metadata:
    name: pod-identity-webhook
  webhooks:
  - admissionReviewVersions:
    - v1beta1
    failurePolicy: Ignore
    matchPolicy: Exact
    name: iam-for-pods.amazonaws.com
     ....
    sideEffects: Unknown
```

Unfortunately on EKS it seems that the `pod-identity-webhook` has not been updated to support dry run request (tested on EKS 1.16 ).

Regarding GKE it seems that built-in admission webhooks are compatible since 1.15.

I will update the e2e test to ensure that it is only run on GKE for K8S > 1.15.

TL;DR: even if dry run requests have been promoted to beta in K8S 1.13 the admission webhooks are not all compatible with that feature 😢

File: test/e2e/es/podtemplate_test.go (hunk 3 above)
Comment (barkbay) [reply test/e2e/es/podtemplate_test.go:83]: Agreed but this file is formatted with `go fmt` 🤷‍♂️

File: pkg/controller/elasticsearch/sset/validation.go (hunk 4 above)
Comment (barkbay) [reply pkg/controller/elasticsearch/sset/validation.go:71]: I think that if we reach a timeout here it is either:
* a transient error, in this case it is fair to try again in the next reconciliation loop
* a permanent error, in which case the StatefulSet update/creation that should happen right after will also fail

The error is already reported in the controller log, I'm not sure we should duplicate it.
